### PR TITLE
Add in-context Email Protection sign-up and sign-in flow

### PR DESCRIPTION
**Pull request by CDRussell**
Task/Issue URL: https://app.asana.com/0/488551667048375/1205275665505559/f 

### Description
Adds support for letting users sign up/in to Email Protection directly from a form with an email address input.

Note restrictions on when the in-context prompt will **not** be shown:
- locale is not English
- user installed the app more than 21 days ago
- user said to "never ask again"
- user is already signed into Email Protection

Translations will be done in a follow-up task: [Translations for native in-context email protection copy](https://app.asana.com/0/1205707833624110/1205707833624110/f)

#### Play build types
Play build respects the remote config flag setting for whether we offer up in-context prompts or not. Remote config is currently disabled (will be enabled soon)

- [x] Clean Install the app using `play` build type
- [x] Visit https://fill.dev/form/registration-email
- [x] Verify you do **not** see the grey Dax icon in the email field
- [x] Visit Settings->Logins, and enable autofill toggle
- [x] Return to the webpage and refresh. 
- [x] Verify you still do **not** see the grey Dax icon in the email field (you should see password generation icons though)

**Testing when remote config is enabled**
- [x] Hack the remote config location (in `PrivacyConfigService`) to use an alternative JSON where feature is enabled (e.g., https://jsonblob.com/api/1159434123311177728) 
- [x] Run the `play` build again, and give it time to fetch the remote config
- [x] Visit https://fill.dev/form/registration-email again
- [x] Verify you **do** see the grey Dax icon on the email field

### Testing the in-context sign up flow
- [x] Visit https://fill.dev/form/registration-email and tap the grey Dax icon, and choose primary `Protect My Email` CTA
- [x] Choose `Get Started` and follow the flow for signing up for a new address
- [x] Verify X button behaviours: 
    - [x] At the screen where you can "Choose your Duck Address" (and onwards), tapping the X button will prompt the user to confirm they want to exit. 
    - [x] Tapping the X button in any screen before that should exit without a confirmation prompt
- [x] Complete the sign up flow until you see the "Try It Now" button. Tap that, and verify you see the "choose duck address" bottom sheet prompt and that it works if you select an option.

### Testing the in-context sign in flow (manually typing code)
- [x] Log out of Email Protection
- [x] Visit https://fill.dev/form/registration-email and tap the grey Dax icon, and choose `I Already Have a Duck Address` button
- [x] Enter your duck address, then get your verification code in your email. Come back to the app and enter the code
- [x] Verify in-context flow immediately ends, you are signed in, and you see the "choose duck address" bottom sheet prompt

### Testing the in-context sign in flow (opening the email verification link in our app)
- [x] Log out of Email Protection
- [x] Set our app as default browser (or choose to open the email verification link with us when prompted)
- [x] Visit https://fill.dev/form/registration-email and tap the grey Dax icon, and choose `I Already Have a Duck Address` button
- [x] Enter your duck address, then click the verification link in your email so it opens our app
- [x] Verify in-context flow immediately ends, you are signed in, and you see the "choose duck address" bottom sheet prompt

### Testing the in-context sign in flow (opening the email verification link in our app but user exited in-context view)
- [x] Log out of Email Protection
- [x] Set our app as default browser (or choose to open the email verification link with us when prompted)
- [x] Visit https://fill.dev/form/registration-email and tap the grey Dax icon, and choose `I Already Have a Duck Address` button
- [x] Enter your duck address so that a verification email is sent; exit the in-context flow here
- [x] Tap the verification link in your email and let it launch in our app
- [x] Verify in-context view is not shown
- [x] Verify a new tab with the email settings is shown, and that you're logged in ok

### Testing locale restriction
- [x] Log out of Email Protection
- [x] Set device locale to something that is non-English
- [x] Visit https://fill.dev/form/registration-email and verify grey Dax icon is **not** shown
- [x] (Return device locale to English so it doesn't interfere with the rest of the tests)

### Testing time-since-install restriction
- [x] Log out of Email Protection
- [x] Set date on the device to be 30 days into the future
- [x] Visit https://fill.dev/form/registration-email and verify grey Dax icon is **not** shown

### Testing the "Don't Show Again" button
- [x] Log out of Email Protection
- [x] Visit https://fill.dev/form/registration-email and tap the grey Dax icon, and choose "Don't Show Again"
- [x] Verify the grey Dax icon disappears
- [x] Refresh the page, and verify it does not show up again (ever)

### UI changes
form-combined

Screenshot_20231012_105239

Screenshot_20231012_105251

### CODE REVIEW

**Comment by CDRussell** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3619** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3619" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3619?utm_source=stack-comment).

File: app/build.gradle [hunk 1]
```diff
@@ -174,6 +174,7 @@ dependencies {
     implementation project(':autofill-api')
     implementation project(':autofill-impl')
     implementation project(':autofill-store')
+    internalImplementation project(':autofill-internal')
```
Comment (CDRussell) [inline app/build.gradle:177]: Adds some dev settings for autofill and email protection (like clearing the "never ask again" choice)

File: package.json [hunk 2]
```diff
@@ -15,7 +15,7 @@
   },
   "dependencies": {
     "@duckduckgo/autoconsent": "^5.1.0",
-    "@duckduckgo/autofill": "github:duckduckgo/duckduckgo-autofill#8.4.2",
+    "@duckduckgo/autofill": "github:duckduckgo/duckduckgo-autofill#abrown/android-incontext-email-promotion",
```
Comment (CDRussell) [inline package.json:18]: This will be changed to a stable release version before merging PR (waiting on it being made currently)

**Review by marcosholgado — APPROVED** [review]
LGTM, left a few comments.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/jsbridge/request/AutofillRequestParser.kt [hunk 3]
```diff
@@ -29,7 +29,7 @@ interface AutofillRequestParser {
     suspend fun parseStoreFormDataRequest(request: String): AutofillStoreFormDataRequest
 }
 
-@ContributesBinding(FragmentScope::class)
+@ContributesBinding(AppScope::class)
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/jsbridge/request/AutofillRequestParser.kt:32]: https://github.com/duckduckgo/Android/pull/3619#discussion_r1363944039

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/jsbridge/AutofillMessagePoster.kt [hunk 4]
```diff
@@ -36,7 +36,7 @@ interface AutofillMessagePoster {
     )
 }
 
-@ContributesBinding(FragmentScope::class)
+@ContributesBinding(AppScope::class)
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/jsbridge/AutofillMessagePoster.kt:39]: https://github.com/duckduckgo/Android/pull/3619#discussion_r1363944039

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/ui/credential/passwordgeneration/AutogeneratedPasswordEventResolver.kt [hunk 5]
```diff
@@ -39,7 +39,7 @@ sealed interface Actions {
     object DiscardAutoLoginId : Actions
 }
 
-@ContributesBinding(FragmentScope::class)
+@ContributesBinding(AppScope::class)
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/ui/credential/passwordgeneration/AutogeneratedPasswordEventResolver.kt:42]: https://github.com/duckduckgo/Android/pull/3619#discussion_r1363944039

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillJavascriptInterface.kt [hunk 6]
```diff
@@ -69,17 +76,24 @@ interface AutofillJavascriptInterface {
...
     companion object {
         const val INTERFACE_NAME = "BrowserAutofill"
     }
+
+    @JavascriptInterface
+    fun closeEmailProtectionTab(data: String)
 }
 
-@ContributesBinding(FragmentScope::class)
+@ContributesBinding(AppScope::class)
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillJavascriptInterface.kt:96]: https://github.com/duckduckgo/Android/pull/3619#discussion_r1363944039

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/email/incontext/verifier/RealEmailProtectionLinkVerifier.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package com.duckduckgo.autofill.impl.email.incontext.verifier
+
+import com.duckduckgo.autofill.api.emailprotection.EmailProtectionLinkVerifier
+import com.duckduckgo.di.scopes.AppScope
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+
+@ContributesBinding(AppScope::class)
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/email/incontext/verifier/RealEmailProtectionLinkVerifier.kt:24]: https://github.com/duckduckgo/Android/pull/3619#discussion_r1363944039

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/email/incontext/availability/EmailProtectionInContextAvailabilityRules.kt [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+import com.squareup.anvil.annotations.ContributesBinding
+import java.util.*
+import javax.inject.Inject
+import kotlinx.coroutines.withContext
+
+interface EmailProtectionInContextAvailabilityRules {
+    suspend fun permittedToShow(url: String): Boolean
+}
+
+@ContributesBinding(AppScope::class)
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/email/incontext/availability/EmailProtectionInContextAvailabilityRules.kt:35]: https://github.com/duckduckgo/Android/pull/3619#discussion_r1363944039

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/email/incontext/availability/EmailProtectionInContextRecentInstallChecker.kt [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import com.duckduckgo.di.scopes.AppScope
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+import kotlinx.coroutines.withContext
+
+interface EmailProtectionInContextRecentInstallChecker {
+    suspend fun isRecentInstall(): Boolean
+}
+
+@ContributesBinding(AppScope::class)
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/email/incontext/availability/EmailProtectionInContextRecentInstallChecker.kt:31]: https://github.com/duckduckgo/Android/pull/3619#discussion_r1363944039

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/email/incontext/ResultHandlerInContextEmailProtectionPrompt.kt [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+import com.duckduckgo.autofill.impl.email.incontext.store.EmailProtectionInContextDataStore
+import com.duckduckgo.di.scopes.AppScope
+import com.squareup.anvil.annotations.ContributesMultibinding
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+
+@ContributesMultibinding(AppScope::class)
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/email/incontext/ResultHandlerInContextEmailProtectionPrompt.kt:42]: https://github.com/duckduckgo/Android/pull/3619#discussion_r1363944039

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/jsbridge/response/AutofillResponseWriter.kt [hunk 11]
```diff
@@ -27,15 +28,19 @@ interface AutofillResponseWriter {
     fun generateEmptyResponseGetAutofillData(): String
     fun generateResponseForAcceptingGeneratedPassword(): String
     fun generateResponseForRejectingGeneratedPassword(): String
+    fun generateResponseForEmailProtectionInContextSignup(installedRecently: Boolean, permanentlyDismissedAtTimestamp: Long?): String
+    fun generateResponseForEmailProtectionEndOfFlow(isSignedIn: Boolean): String
 }
 
-@ContributesBinding(FragmentScope::class)
+@ContributesBinding(AppScope::class)
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/jsbridge/response/AutofillResponseWriter.kt:35]: Tried with `ActivityScope`, individually, and in various combinations but each time it wouldn't compile (I think that there is something higher up demanding `AppScope` in each chain)

**Pull request merged**